### PR TITLE
Add Access-Control-Expose-Headers

### LINKED ISSUE #12424: Access-Control-Allow-Headers in API
- Gitea version (or commit ref): v1.12 and HEAD branch
- Git version: 2.20.1
- Operating system: Debian Buster 10.4
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No: Seems broken, throws 500 error
  - [ ] Not relevant
- Log gist:

## Description

I already postet some information here:
https://github.com/go-gitea/gitea/pull/6408

I try to give a summary.

I want to access the list of all commits in a given repo in a Vue app using axios.

My Gitea server works fine, accessing the info via curl works, even in Chrome I get the expected headers (see screenshot), but the cors settings prevent the Javascript, or especially axios to access the data.

Here are some background infos:
https://stackoverflow.com/questions/37897523/axios-get-access-to-response-header-fields

So, I need to tell Gitea to send X-Total in the Access-Control-Allow-Headers, I already glimpsed through the code, tried to change it myself, only to find out later that it was the part related to the general http stuff, not the API.

That is part of my code:

`			async fetchData() {
				this.isLoading = true;
				axios.get(`https://git.24unix.net:3000/api/v1/repos/24unix.net/yyyy/commits?page=${this.currentPage}&limit=10`,
					{
						headers: {
							Authorization: 'token ' + "bfebd3123cyyyyyyyyyyyyyyy34abe46f1a",
						}
					})
				.then((response) => {
`

Response headers just contain the Content-Type.

Chrome receives the relevant headers:

image

### PR DESCRIPTION
**Pull request by zeripath**
Fix #12424

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by silverwind** [comment]
Should probably only set them if `cors.ENABLED` is set.

**Comment by zeripath** [comment]
Does it matter? I mean it's still semantically useful?

**Comment by silverwind** [comment]
It shouldn't really matter but CORS-specific headers do not convey any meaning when not using CORS so I see them mostly as garbage information in such a scenario.

**Comment by silverwind** [comment]
I guess ideally we should check for the existance of the `Origin` request header to detect CORS requests and then set the `Access-Control-*` headers accordingly when its present. That way it will even work when CORS headers were added by a reverse proxy (and not by ourselves).

**Comment by zeripath** [comment]
I think that detecting whether CORS is being set etc through use of the Origin header is likely to be error prone - I can't see a problem with just emitting these in any case.

Do we really need to add the additional x lines of code to detect if we think CORS is enabled and Origin is set correctly etc. (with all the risk of getting that wrong) in these 14 places to determine whether we strictly need to emit this header that does no harm if it's not strictly needed but does no harm if it's emitted otherwise?

File: modules/context/context.go [hunk 1]
```diff
@@ -222,6 +222,7 @@ func (ctx *Context) ServeContent(name string, r io.ReadSeeker, params ...interfa
 	ctx.Resp.Header().Set("Expires", "0")
 	ctx.Resp.Header().Set("Cache-Control", "must-revalidate")
 	ctx.Resp.Header().Set("Pragma", "public")
+	ctx.Header().Set("Access-Control-Expose-Headers", "Content-Disposition")
```
Comment (silverwind) [inline modules/context/context.go]: Is there a difference between `ctx.Resp.Header` and `ctx.Header`? Should we really mix them like here (and a few more places in this PR).

**Comment by silverwind** [comment]
`Origin` will be present on `POST` requests so it'll have some false positives. I guess I don't really care if they are always set.

File: modules/context/context.go (hunk 1 above)
Comment (zeripath) [reply modules/context/context.go]:
```suggestion
	ctx.Resp.Header().Set("Access-Control-Expose-Headers", "Content-Disposition")
```

File: modules/lfs/server.go [hunk 2]
```diff
@@ -204,6 +205,7 @@ func getContentHandler(ctx *context.Context) {
 		decodedFilename, err := base64.RawURLEncoding.DecodeString(filename)
 		if err == nil {
 			ctx.Resp.Header().Set("Content-Disposition", "attachment; filename=\""+string(decodedFilename)+"\"")
+			ctx.Header().Set("Access-Control-Expose-Headers", "Content-Disposition")
```
Comment (zeripath) [inline modules/lfs/server.go]:
```suggestion
			ctx.Resp.Header().Set("Access-Control-Expose-Headers", "Content-Disposition")
```

File: modules/context/context.go (hunk 1 above)
Comment (zeripath) [reply modules/context/context.go]: I don't think there is a difference, no. But for consistencies sake I'll change it.

File: modules/lfs/server.go [hunk 3]
```diff
@@ -183,6 +183,7 @@ func getContentHandler(ctx *context.Context) {
 			}
 
 			ctx.Resp.Header().Set("Content-Range", fmt.Sprintf("bytes %d-%d/%d", fromByte, toByte, meta.Size-fromByte))
+			ctx.Header().Set("Access-Control-Expose-Headers", "Content-Range")
```
Comment (silverwind) [inline modules/lfs/server.go]:
```suggestion
			ctx.Resp.Header().Set("Access-Control-Expose-Headers", "Content-Range")
```

File: routers/api/v1/repo/commits.go [hunk 4]
```diff
@@ -212,6 +212,7 @@ func GetAllCommits(ctx *context.APIContext) {
 	ctx.SetLinkHeader(int(commitsCountTotal), listOptions.PageSize)
 	ctx.Header().Set("X-Total-Count", fmt.Sprintf("%d", commitsCountTotal))
 
+	ctx.Header().Set("Access-Control-Expose-Headers", "X-Total-Count, X-PerPage, X-Total, X-PageCount, X-HasMore, Link")
```
Comment (silverwind) [inline routers/api/v1/repo/commits.go]:
```suggestion
	ctx.Header().Set("Access-Control-Expose-Headers", "X-Total-Count, X-PerPage, X-Total, X-PageCount, X-HasMore, Link")

```

**Review by silverwind — APPROVED** [review]

**Review by techknowlogick — APPROVED** [review]

**Pull request merged**